File: frontend/src/lib/scenarioDialogue.ts
/**
 * scenarioDialogue — 情境對話面板 state 與後端 payload 的對映層（Issue #1013）。
 *
 * 對應後端 `backend/utils/scenario_dialogue.py`（那裡是儲存語意的唯一判定處），
 * 這裡只做「面板 camelCase ↔ API snake_case」的搬運，不做任何值的推導。
 *
 * ## 為什麼獨立成一個模組
 *
 * 面板（`ScenarioDialoguePanel`）負責畫面、呼叫端（`TeacherTemplatePrograms`）負責
 * 打 API，兩邊都需要同一份對映；寫在任一邊都會讓另一邊複製一份，而這份對映有兩個
 * 一旦寫歪就很難察覺的規則（見下），複製等於埋兩顆同樣的雷。
 *
 * ## 規則一：`null` 不可被壓平
 *
 * `tenseOverride` / `voiceOverride` 有三種狀態，畫面上後兩種看起來都是空的，行為卻不同：
 *
 * | 值 | 意思 | 老師之後改整體評分標準 |
 * |---|---|---|
 * | `null` | 沿用整體 | **本題跟著變** |
 * | `{ time: "", aspect: "" }` / `""` | 本題已脫鉤，明確不指定 | 本題不受影響 |
 * | `{ time: "past", aspect: "simple" }` / `"passive"` | 本題自訂 | 本題不受影響 |
 *
 * 所以這裡一律用 `== null` 判斷，**禁止**用 `||` / `if (!x)` 把 `null` 與空值折疊
 * 成同一個值 —— 折疊後畫面完全看不出差別，要等 AI 用錯時態評分才會發現。
 *
 * ## 規則二：`referenceAnswer` 只走老師端
 *
 * 參考答案只給 AI 當語言特徵對照，學生端看不到。本模組服務的是老師端（出題／編輯），
 * 所以會帶上它；學生端的序列化在後端 `public_item_view`，不要在前端另做一套。
 */

/** 時態＝時間 × 動貌，兩者都選才成立（避免「過去」但沒說哪一種的半套條件） */
export interface TenseSetting {
  /** `present` / `past` / `future`，或空字串代表不指定 */
  time: string;
  /** `simple` / `progressive` / `perfect` / `perfectProgressive`，或空字串 */
  aspect: string;
}

/** 時態／語態都沒指定時的值。整份設定沒有上層可沿用，缺值時落在這裡而不是 `null` */
export const EMPTY_TENSE: TenseSetting = { time: "", aspect: "" };

/** 面板的一列題目（`ScenarioDialoguePanel` 的 state 單位） */
export interface ScenarioDialogueRow {
  /** 前端用的列 key（拖曳排序），與資料庫無關 */
  id: string;
  /**
   * 這一列在資料庫的 `content_items.id`；新題為 `null`。
   *
   * **編輯既有內容時一定要帶回後端**：後端是 id-based 原地更新（#861），payload 沒帶
   * id 就會把整批題目刪掉重建，學生作答（practice_answers / student_item_progress）
   * 綁的是舊 id，會隨 FK CASCADE 一起消失、批改分數歸零。
   */
  contentItemId: number | null;
  /** 口說題目本文（老師可自行編輯） */
  question: string;
  /** 輔助語言翻譯，語言由整份設定統一決定 */
  translation: string;
  /**
   * 時態覆寫。`null` = 沿用「整份設定」的整體評分標準，會跟著整體變動；
   * 一旦老師動過就固定成本題自訂，之後改整體不再影響它（可用 ↺ 復原成沿用）。
   */
  tenseOverride: TenseSetting | null;
  /** 語態覆寫。語意同 tenseOverride */
  voiceOverride: string | null;
  /** 必用字詞 —— 逐題獨立，不繼承（每題要練的單字本來就不同） */
  keywords: string[];
  /**
   * 參考答案 —— 只給 AI 當評分對照，**學生端完全看不到**。
   *
   * 定位是「示範回答」而非唯一正解：口說同一題每個學生的內容本來就不同，
   * 送 AI 時比對的是**結構與語言特徵**（時態、句型、用字水準、資訊完整度），
   * 絕不可拿來做逐字比對，否則所有與範例不同的答案都會被誤判。
   * 依 #864 規格 3-3，學生答案需修正時，它也是「建議的答案」的依據。
   */
  referenceAnswer: string;
  /** 本題額外說明（選填）。與全份共用的作答指引一起給 AI 與學生看 */
  rubricNote: string;
  /**
   * 情境圖片的生成 prompt。
   *
   * UI 不顯示也不讓老師編輯（圖片改成「AI 生成」或手動上傳二擇一），但**要存**：
   * #1013 定案讓後端保留這個欄位，因為 AI 產題預期會連 prompt 一起回傳給後續
   * 重新生圖用，前端只是搬運，不解讀內容。
   */
  imagePrompt: string;
  /** 已生成的情境圖片；null = 尚未生成 */
  imageUrl: string | null;
  /** 題目語音；null = 尚未生成 */
  audioUrl: string | null;
  /**
   * 內容被整批換掉的次數（目前只有「重新生成這一題」會加）。
   *
   * SortableRow 有兩個從 props 取一次初值的本地狀態：關鍵字草稿與「參考答案」
   * 展開與否。重新生成沿用同一個 row id，React 不會重新掛載，這兩個本地狀態
   * 就會停在舊值 —— 關鍵字輸入框顯示舊字，老師只要 focus 再 blur，onBlur 的
   * commitKeywords 就用舊草稿把新關鍵字蓋回去，而且沒有任何提示。
   *
   * 所以 render 時把 key 設成 `id:revision`：id 保持穩定給拖曳排序用，
   * revision 一變就重新掛載，本地狀態自然跟著重新取一次初值。
   *
   * 純前端狀態，不進 API payload。
   */
  revision: number;
}

export interface ScenarioTTSSettings {
  accent: string;
  gender: string;
  speed: string;
}

/** 面板 `onSave` 交出來的整份資料 */
export interface ScenarioSaveInput {
  title: string;
  rows: ScenarioDialogueRow[];
  scenarioContent: string;
  questionLevel: string;
  globalRubric: string;
  globalTense: TenseSetting;
  globalVoice: string;
  translateLanguage: string;
  ttsSettings: ScenarioTTSSettings;
}

/** 寫進 `contents.scenario_settings` 的整份設定 */
export interface ScenarioSettingsPayload {
  scenario_content: string;
  question_level: string;
  global_rubric: string;
  global_tense: TenseSetting;
  global_voice: string;
  translate_language: string;
  tts_settings: ScenarioTTSSettings | null;
}

/** 寫進 `content_items.item_metadata["scenario_dialogue"]` 的逐題區塊 */
export interface ScenarioItemBlock {
  tense_override: TenseSetting | null;
  voice_override: string | null;
  keywords: string[];
  reference_answer: string;
  rubric_note: string;
  image_prompt: string;
}

/** 送給 create / update content 的單題 payload */
export interface ScenarioItemPayload {
  /** 既有題目才有；新題不帶（後端據此決定原地更新或新增，#861） */
  id?: number;
  text: string;
  translation: string;
  image_url: string | null;
  audio_url: string | null;
  scenario_dialogue: ScenarioItemBlock;
}

export interface ScenarioSavePayload {
  scenario_settings: ScenarioSettingsPayload;
  items: ScenarioItemPayload[];
}

/** `GET /api/teachers/contents/{id}` 之中本題型用得到的部分 */
export interface ScenarioContentDetail {
  id?: number;
  title?: string | null;
  scenario_settings?: Partial<ScenarioSettingsPayload> | null;
  items?: Array<{
    id?: number;
    text?: string | null;
    translation?: string | null;
    image_url?: string | null;
    audio_url?: string | null;
    scenario_dialogue?: Partial<ScenarioItemBlock> | null;
  }> | null;
}

/** 編輯模式回填面板用的初始 state */
export interface ScenarioDialogueInitialState {
  title: string;
  scenarioContent: string;
  questionLevel: string;
  globalRubric: string;
  globalTense: TenseSetting;
  globalVoice: string;
  translateLanguage: string;
  /** `null` = 這份內容沒存過 TTS 設定 → 呼叫端沿用自己的預設值 */
  ttsSettings: ScenarioTTSSettings | null;
  rows: ScenarioDialogueRow[];
}

const text = (value: unknown): string =>
  typeof value === "string" ? value : "";

let rowSeq = 0;
/** 拖曳排序要穩定的 key，時間戳加序號足以避免同一毫秒內連續建列撞號 */
export const nextRowId = () => `sd-${Date.now()}-${rowSeq++}`;

/** 建一列空白題目。面板一開就給一張空卡，老師可以直接打字不必先產題 */
export const createScenarioRow = (
  overrides: Partial<ScenarioDialogueRow> = {},
): ScenarioDialogueRow => ({
  id: nextRowId(),
  contentItemId: null,
  question: "",
  translation: "",
  tenseOverride: null,
  voiceOverride: null,
  keywords: [],
  referenceAnswer: "",
  rubricNote: "",
  imagePrompt: "",
  imageUrl: null,
  audioUrl: null,
  revision: 0,
  ...overrides,
});

/**
 * 面板翻譯語言下拉的選項值，與 `ScenarioDialoguePanel` 的 `TRANSLATION_LANGUAGES`
 * 一一對應（改一邊要改另一邊）。`other` 代表「語言名字在旁邊那個自訂欄位裡」。
 */
export const TRANSLATION_LANGUAGE_VALUES = [
  "chinese",
  "japanese",
  "korean",
  "other",
] as const;

/**
 * 面板的（下拉選擇 + 自訂欄位）→ 存進 `scenario_settings.translate_language` 的單一值。
 *
 * 選「其他」時存的是老師真的打的語言名字，**不是**字面的 `"other"` —— 存 `"other"`
 * 等於把語言丟掉：重開編輯還原不出來，送去翻譯的一方也不知道要翻成什麼，而且全程
 * 沒有任何錯誤訊息（PR #1016 review）。選了「其他」卻沒打字則退回 `"other"`，
 * 至少把「老師選過其他」這件事留住。
 */
export function toStoredTranslateLanguage(
  selected: string,
  custom: string,
): string {
  if (selected !== "other") return text(selected);
  return text(custom).trim() || "other";
}

/** 上面那個的反向：存下來的單一值 → 面板的（下拉選擇 + 自訂欄位）。 */
export function fromStoredTranslateLanguage(stored: string): {
  selected: string;
  custom: string;
} {
  const value = text(stored);
  if (!value) return { selected: "", custom: "" };
  // 認得的選項值原樣還原；其餘一律視為自訂語言
  return (TRANSLATION_LANGUAGE_VALUES as readonly string[]).includes(value)
    ? { selected: value, custom: "" }
    : { selected: "other", custom: value };
}

/** 時態正規化。`null`／`undefined` 一律保持 `null`（＝沿用整體），不可填成空物件 */
const toTensePayload = (tense: TenseSetting | null | undefined) =>
  tense == null ? null : { time: text(tense.time), aspect: text(tense.aspect) };

/**
 * 面板 state → API payload。
 *
 * 沒打題目的列會被略過（面板預設就有一張空白卡），題數上下限由面板與後端各擋一次，
 * 這裡不重複判斷 —— 對映層只負責換形狀。
 */
export function toScenarioSavePayload(
  input: ScenarioSaveInput,
): ScenarioSavePayload {
  const scenario_settings: ScenarioSettingsPayload = {
    // 情境內容空字串是合法值（老師只給標題、自己出題），不要當成缺值
    scenario_content: text(input.scenarioContent),
    question_level: text(input.questionLevel),
    global_rubric: text(input.globalRubric),
    global_tense: toTensePayload(input.globalTense) ?? { ...EMPTY_TENSE },
    global_voice: text(input.globalVoice),
    translate_language: text(input.translateLanguage),
    tts_settings: input.ttsSettings
      ? {
          accent: text(input.ttsSettings.accent),
          gender: text(input.ttsSettings.gender),
          speed: text(input.ttsSettings.speed),
        }
      : null,
  };

  const items = (input.rows || [])
    .filter((row) => text(row.question).trim())
    .map((row) => ({
      // 既有題目帶上 DB id，後端才會原地更新而不是刪掉重建（#861）
      ...(row.contentItemId == null ? {} : { id: row.contentItemId }),
      // 題目本文／翻譯／圖片／音檔沿用既有欄位，不塞進 scenario_dialogue
      text: text(row.question).trim(),
      translation: text(row.translation).trim(),
      image_url: row.imageUrl ?? null,
      audio_url: row.audioUrl ?? null,
      scenario_dialogue: {
        // 兩個 override 一律保留 null／值的差別，這裡不可以用 `||` 折疊
        tense_override: toTensePayload(row.tenseOverride),
        voice_override:
          row.voiceOverride == null ? null : text(row.voiceOverride),
        keywords: (row.keywords || [])
          .map((k) => text(k).trim())
          .filter(Boolean),
        reference_answer: text(row.referenceAnswer),
        rubric_note: text(row.rubricNote),
        image_prompt: text(row.imagePrompt),
      },
    }));

  return { scenario_settings, items };
}

/**
 * API 回傳 → 面板初始 state（編輯模式）。
 *
 * 舊資料（或非本題型轉過來的內容）可能完全沒有 `scenario_settings` /
 * `scenario_dialogue`，一律落成安全預設而不是丟錯 —— 面板打不開的話老師連改都沒得改。
 * 沒存過覆寫的題目視為「沿用整體」（`null`），與面板新建列的預設一致。
 */
export function fromScenarioContentDetail(
  detail: ScenarioContentDetail,
): ScenarioDialogueInitialState {
  const settings = detail.scenario_settings || {};
  const rawTense = settings.global_tense;
  const tts = settings.tts_settings;

  return {
    title: text(detail.title),
    scenarioContent: text(settings.scenario_content),
    questionLevel: text(settings.question_level),
    globalRubric: text(settings.global_rubric),
    globalTense: rawTense
      ? { time: text(rawTense.time), aspect: text(rawTense.aspect) }
      : { ...EMPTY_TENSE },
    globalVoice: text(settings.global_voice),
    translateLanguage: text(settings.translate_language),
    ttsSettings: tts
      ? {
          accent: text(tts.accent),
          gender: text(tts.gender),
          speed: text(tts.speed),
        }
      : null,
    rows: (detail.items || []).map((item) => {
      const block = item.scenario_dialogue || {};
      return createScenarioRow({
        contentItemId: typeof item.id === "number" ? item.id : null,
        question: text(item.text),
        translation: text(item.translation),
        // undefined（沒存過）與 null（沿用整體）都落成 null；只有物件才是脫鉤，
        // 且脫鉤後的空物件 { time: "", aspect: "" } 要原樣保留，不可再折回 null
        tenseOverride: block.tense_override
          ? {
              time: text(block.tense_override.time),
              aspect: text(block.tense_override.aspect),
            }
          : null,
        voiceOverride: block.voice_override ?? null,
        keywords: Array.isArray(block.keywords) ? block.keywords.map(text) : [],
        referenceAnswer: text(block.reference_answer),
        rubricNote: text(block.rubric_note),
        imagePrompt: text(block.image_prompt),
        imageUrl: item.image_url ?? null,
        audioUrl: item.audio_url ?? null,
      });
    }),
  };
}

File: backend/utils/scenario_dialogue.py
"""情境對話（Scenario Dialogue）資料正規化 — Issue #1013。

老師端面板 ``frontend/src/components/ScenarioDialoguePanel.tsx`` 比其他題型多出
兩層資料，這個模組是它們「怎麼存」的**唯一**判定處：

* **整份設定** —— 情境內容、題目難度（CEFR）、作答指引、整體評分標準（時態／
  語態）、翻譯語言、TTS 設定。存進 ``contents.scenario_settings``（JSONB）。
* **逐題資料** —— 時態／語態覆寫、必用字詞、參考答案、評分備註、圖片 prompt。
  存進 ``content_items.item_metadata['scenario_dialogue']``。

題目本文 / 翻譯 / 圖片 / 音檔沿用既有欄位（``text`` / ``translation`` /
``image_url`` / ``audio_url``），不另開欄位。

逐題資料放 ``item_metadata`` 而不是新欄位，有兩個理由：

1. 作業副本（``routers/assignments/crud.py``）、即刻練習、教材複製都已經把
   ``item_metadata`` 整包 copy 過去，放這裡等於自動跟著複製，不必再改一輪；
2. ``tense_override`` 需要區分兩種**行為不同**的狀態，只有 JSON 表達得出來（見下）。

tense_override / voice_override 的 null 語意（**不可壓平**）
-----------------------------------------------------------

``None``
    沿用整份設定的 ``global_tense`` / ``global_voice``。之後老師改整體，**本題跟著變**。

``{"time": "", "aspect": ""}`` / ``""``
    本題已脫鉤，明確「不指定」。之後改整體，**本題不受影響**。

兩者在 UI 上看起來都是空的，但行為完全不同。因此本模組一律用 ``is None`` 判斷，
**禁止**用 truthiness（``or`` / ``if not x``）把兩者壓成同一個值。生效值請一律走
:func:`resolve_effective_tense` / :func:`resolve_effective_voice`，不要在呼叫端自己寫。

reference_answer 不得外流到學生端
---------------------------------

``reference_answer`` 只給 AI 評分當**語言特徵**對照（時態、句型、用字水準、資訊
完整度），學生端絕對看不到，而且**不可拿來逐字比對** —— 口說同一題每個學生講的
內容本來就不同，逐字比對會把所有與範例不同的答案誤判成錯。任何送往學生端的序列化
都必須走 :func:`public_item_view`，它會把 ``reference_answer`` 與 ``image_prompt``
（內部用的生成 prompt）拿掉。

穩定代碼
--------

時態／語態存的是穩定代碼（``past`` / ``simple`` / ``perfectProgressive`` /
``active`` …），不是中文。代碼與前端 ``TENSE_TIMES`` / ``TENSE_ASPECTS`` /
``VOICES`` 一一對應，老師切換介面語言不會改變存進資料庫的值。未知代碼一律擋下
（丟 ``ValueError``），避免前端／AI 產題送進髒資料後才在評分階段爆掉。
"""
from typing import Any, Dict, List, Optional

# item_metadata 底下的命名空間 key
SCENARIO_ITEM_KEY = "scenario_dialogue"

# 一份情境對話的題數上下限（#864；前端 MIN_ITEMS / MAX_ITEMS 同值）
MIN_ITEMS = 3
MAX_ITEMS = 10

# ---- 穩定代碼（與 ScenarioDialoguePanel.tsx 的常數一一對應）----
TENSE_TIMES = frozenset({"present", "past", "future"})
TENSE_ASPECTS = frozenset({"simple", "progressive", "perfect", "perfectProgressive"})
VOICES = frozenset({"active", "passive"})
CEFR_LEVELS = frozenset({"A1", "A2", "B1", "B2", "C1", "C2"})

# 「不指定」的時態 —— 整份設定沒有「沿用」的上層，所以缺值時落在這裡而不是 None
EMPTY_TENSE: Dict[str, str] = {"time": "", "aspect": ""}


def _text(value: Any) -> str:
    """任何缺值都收斂成空字串（空字串在本題型是合法值，不是「未設定」）。"""
    if value is None:
        return ""
    if not isinstance(value, str):
        raise ValueError(f"預期字串，收到 {type(value).__name__}")
    return value


def _code(value: Any, allowed: frozenset, field: str) -> str:
    """驗證穩定代碼。空字串 = 不指定（合法）；未知代碼直接擋下。"""
    text = _text(value)
    if text == "":
        return ""
    if text not in allowed:
        raise ValueError(f"{field} 不是合法代碼：{text!r}（可用：{sorted(allowed)}）")
    return text


def normalize_tense(raw: Any, field: str = "tense") -> Optional[Dict[str, str]]:
    """時態正規化。

    ``None`` 原封不動回傳 ``None``（逐題＝沿用整體；整體＝呼叫端自行補 EMPTY_TENSE）。
    """
    if raw is None:
        return None
    if not isinstance(raw, dict):
        raise ValueError(f"{field} 需為物件或 null，收到 {type(raw).__name__}")
    return {
        "time": _code(raw.get("time"), TENSE_TIMES, f"{field}.time"),
        "aspect": _code(raw.get("aspect"), TENSE_ASPECTS, f"{field}.aspect"),
    }


def normalize_voice(raw: Any, field: str = "voice") -> Optional[str]:
    """語態正規化。``None`` 保持 ``None``（沿用整體），``""`` 是「明確不指定」。"""
    if raw is None:
        return None
    return _code(raw, VOICES, field)


def _normalize_keywords(raw: Any) -> List[str]:
    """必用字詞 —— 逐題獨立、不繼承整體。去掉空白項，保留老師輸入的順序。"""
    if raw is None:
        return []
    if not isinstance(raw, list):
        raise ValueError(f"keywords 需為陣列，收到 {type(raw).__name__}")
    return [w.strip() for w in (_text(x) for x in raw) if w.strip()]


def _normalize_tts_settings(raw: Any) -> Optional[Dict[str, str]]:
    """TTS 設定（accent / gender / speed）。沒帶就是沒帶，不硬塞預設值。"""
    if raw is None:
        return None
    if not isinstance(raw, dict):
        raise ValueError(f"tts_settings 需為物件或 null，收到 {type(raw).__name__}")
    return {
        "accent": _text(raw.get("accent")),
        "gender": _text(raw.get("gender")),
        "speed": _text(raw.get("speed")),
    }


def normalize_settings(raw: Any) -> Optional[Dict[str, Any]]:
    """整份設定正規化，回傳可直接寫進 ``contents.scenario_settings`` 的 dict。

    ``None`` 代表「這次請求沒帶設定」→ 呼叫端不要動既有值。
    """
    if raw is None:
        return None
    if not isinstance(raw, dict):
        raise ValueError(f"scenario_settings 需為物件，收到 {type(raw).__name__}")

    # 整體評分標準沒有「沿用上層」這回事，所以 None 落成 EMPTY_TENSE 而非保留 None
    global_tense = normalize_tense(raw.get("global_tense"), "global_tense")
    global_voice = normalize_voice(raw.get("global_voice"), "global_voice")

    return {
        # 情境內容：空字串合法（老師只給標題、自己出題也存得起來）
        "scenario_content": _text(raw.get("scenario_content")),
        "question_level": _code(
            raw.get("question_level"), CEFR_LEVELS, "question_level"
        ),
        "global_rubric": _text(raw.get("global_rubric")),
        "global_tense": dict(EMPTY_TENSE) if global_tense is None else global_tense,
        "global_voice": "" if global_voice is None else global_voice,
        "translate_language": _text(raw.get("translate_language")),
        "tts_settings": _normalize_tts_settings(raw.get("tts_settings")),
    }


def normalize_item(raw: Any) -> Dict[str, Any]:
    """逐題資料正規化，回傳寫進 ``item_metadata['scenario_dialogue']`` 的區塊。"""
    if not isinstance(raw, dict):
        raise ValueError(f"{SCENARIO_ITEM_KEY} 需為物件，收到 {type(raw).__name__}")
    return {
        # 兩個 override 一律保留 None／值的差別，這裡不可以用 `or` 折疊
        "tense_override": normalize_tense(raw.get("tense_override"), "tense_override"),
        "voice_override": normalize_voice(raw.get("voice_override"), "voice_override"),
        "keywords": _normalize_keywords(raw.get("keywords")),
        # 只給 AI 當語言特徵對照，學生端看不到（見模組說明）
        "reference_answer": _text(raw.get("reference_answer")),
        "rubric_note": _text(raw.get("rubric_note")),
        "image_prompt": _text(raw.get("image_prompt")),
    }


def build_item_metadata_block(
    item_data: Dict[str, Any],
    existing_metadata: Optional[Dict[str, Any]] = None,
) -> Optional[Dict[str, Any]]:
    """從一筆 item payload 取出情境對話區塊，供寫入 ``item_metadata`` 使用。

    payload 沒帶 ``scenario_dialogue``（或帶 ``null``）時**沿用既有值**：次要儲存
    路徑（例如只更新音檔）不見得會帶整包，重建 metadata 時不能把老師填的參考答案
    與覆寫洗掉 —— 同 ``example_sentence_translation_lang`` 的處理方式（#1004）。

    回傳 ``None`` 代表「沒有東西要寫」，呼叫端就不要放這個 key。
    """
    raw = item_data.get(SCENARIO_ITEM_KEY)
    if raw is None:
        if existing_metadata:
            existing = existing_metadata.get(SCENARIO_ITEM_KEY)
            if isinstance(existing, dict):
                return existing
        return None
    return normalize_item(raw)


def read_item_block(
    item_metadata: Optional[Dict[str, Any]]
) -> Optional[Dict[str, Any]]:
    """從 ``item_metadata`` 取出情境對話區塊（沒有就 ``None``）。"""
    if not item_metadata:
        return None
    block = item_metadata.get(SCENARIO_ITEM_KEY)
    return block if isinstance(block, dict) else None


def resolve_effective_tense(
    item_block: Optional[Dict[str, Any]],
    settings: Optional[Dict[str, Any]],
) -> Dict[str, str]:
    """本題實際生效的時態 —— null 語意的唯一落實處。

    ``tense_override is None`` → 取整體；否則取本題自己的值（即使是空的）。
    """
    override = (item_block or {}).get("tense_override")
    if override is None:
        global_tense = (settings or {}).get("global_tense")
        if isinstance(global_tense, dict):
            return {
                "time": global_tense.get("time", ""),
                "aspect": global_tense.get("aspect", ""),
            }
        return dict(EMPTY_TENSE)
    return {"time": override.get("time", ""), "aspect": override.get("aspect", "")}


def resolve_effective_voice(
    item_block: Optional[Dict[str, Any]],
    settings: Optional[Dict[str, Any]],
) -> str:
    """本題實際生效的語態。``voice_override is None`` → 取整體。"""
    override = (item_block or {}).get("voice_override")
    if override is None:
        # 用 is None 而不是 `or ""`：本模組自己的規則就是「不可用 truthiness 折疊」，
        # 今天 global_voice 的合法值恰好只有 active / passive / ""，但哪天多一個
        # falsy 的值，`or` 會把它靜靜地當成沒設定。
        global_voice = (settings or {}).get("global_voice")
        return "" if global_voice is None else global_voice
    return override


def public_item_view(
    item_block: Optional[Dict[str, Any]],
    settings: Optional[Dict[str, Any]],
) -> Dict[str, Any]:
    """學生端安全版：**不含** ``reference_answer`` 與 ``image_prompt``。

    override 已在此解析成生效值，學生端不需要（也不該）自己再做繼承推導。
    任何回傳給學生的端點都必須用這個，不要直接吐 ``item_metadata``。
    """
    return {
        "keywords": list((item_block or {}).get("keywords") or []),
        "rubric_note": (item_block or {}).get("rubric_note", ""),
        "tense": resolve_effective_tense(item_block, settings),
        "voice": resolve_effective_voice(item_block, settings),
    }


def teacher_item_view(
    item_block: Optional[Dict[str, Any]],
    settings: Optional[Dict[str, Any]],
) -> Dict[str, Any]:
    """老師端（批改頁）版本：**含** ``reference_answer``。

    與 :func:`public_item_view` 是一對 —— 兩邊的差別只有「參考答案給不給」，所以放在
    一起，改一邊時另一邊就在眼前。批改端點原本是手寫這個 dict（#1034 review 指出），
    那等於在模組外面複製了一份形狀，而本模組的定位就是這些資料「怎麼存、怎麼給」的
    唯一判定處。

    ``reference_answer`` 是**示範回答不是唯一正解**：口說同一題每個學生講的內容本來
    就不同，老師要看的是語言特徵（時態、句型、用字水準、資訊完整度）而不是逐字比對
    （#864 規格 3-3）。批改 Panel 上那行提示就是在講這件事。
    """
    block = item_block or {}
    return {
        **public_item_view(item_block, settings),
        "reference_answer": block.get("reference_answer", ""),
    }


def public_settings_view(settings: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    """學生端安全版的整份設定：情境內容與作答指引學生看得到，其餘是出題端的事。"""
    settings = settings or {}
    return {
        "scenario_content": settings.get("scenario_content", ""),
        "global_rubric": settings.get("global_rubric", ""),
        "translate_language": settings.get("translate_language", ""),
    }


def copy_settings(value: Any) -> Optional[Dict[str, Any]]:
    """複製整份設定，給「內容副本」路徑使用（作業副本／即刻練習／教材複製）。

    副本一定要帶著整份設定走：逐題的 ``tense_override=None`` 是「沿用整體」，副本沒有
    整體可沿用的話，同一題在副本裡的生效時態就會跟原本不一樣。

    非 dict（含 ``None``、其他題型）一律回 ``None``。淺拷貝即可 —— 這份 dict 只有一層
    巢狀（``global_tense`` / ``tts_settings``），而它們在正規化之後就不再被就地修改，
    每次寫入都是整包換新。
    """
    return dict(value) if isinstance(value, dict) else None


def validate_item_count(count: int) -> None:
    """一份情境對話 3~10 題（#864）。前端擋過一次，後端不能只信前端。"""
    if count < MIN_ITEMS or count > MAX_ITEMS:
        raise ValueError(f"情境對話題數需介於 {MIN_ITEMS}~{MAX_ITEMS} 題，收到 {count} 題")
